### PR TITLE
feat(gateway): add `GatewayInstance` CRD for managing Gateway deployments

### PR DESCRIPTION
**Pull request by michaelbeaumont**
### Summary

This PR introduces the new CRD, the name is still open for opinions :)
(`GatewayDataplane` I think somewhat misses that it's more than just a different type of `Dataplane`)

### Issues resolved

Closes https://github.com/kumahq/kuma/issues/3448

~Blocked by https://github.com/kumahq/kuma/pull/3453~ I think I've got it generating just `GatewayInstance` for now

### CODE REVIEW

File: pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_instance.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	// +required
+	// +kubebuilder:validation:MinLen=1
+	Tags map[string]string `json:"tags,omitempty"`
+
+	// Replicas is the number of dataplane proxy replicas to create. For
+	// now this is a fixed number, but in the future it could be
+	// automatically scaled based on metrics.
+	//
+	// +optional
+	// +kubebuilder:default=1
```
Comment (jpeach) [inline pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_instance.go:37]: Add `Minimum=1`

File: pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_instance.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+// +kubebuilder:object:root=true
+
+// GatewayInstance represents a managed instance of a dataplane proxy for a Kuma
+// Gateway.
+type GatewayInstance struct {
+	kube_meta.TypeMeta   `json:",inline"`
+	kube_meta.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec GatewayInstanceSpec `json:"spec,omitempty"`
+}
```
Comment (jpeach) [inline pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_instance.go:17]: Should add `Status` as a subresource.

Comment (jpeach) [reply pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_instance.go:17]: How about this as a starting proposal?

```Go
type GatewayInstanceStatus struct {
	// LoadBalancer contains the current status of the load-balancer,
	// if one is present.
	// +optional
	LoadBalancer *LoadBalancerStatus `json:"loadBalancer,omitempty"

	// Current service state
	// +optional
	// +patchMergeKey=type
	// +patchStrategy=merge
	// +listType=map
	// +listMapKey=type
	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"
}
```

For Conditions, maybe we can get away with just the conventional `Ready`, which is True if both the deployment and service are present and up to date?

**Review by jpeach — COMMENTED** [review]
Looks good! I can live with `GatewayInstance` if you prefer that name. Should we always install this CRD, or only with some special flag?

File: pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_instance.go (hunk 2 above)
Comment (michaelbeaumont) [reply pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_instance.go:18]: Yeah I think that makes sense.

**Comment by michaelbeaumont** [comment]
> I can live with GatewayInstance if you prefer that name

It's not my favorite name ever but we have time to change it if something better occurs to us.

> Should we always install this CRD, or only with some special flag?

I think we should always install it if gateway support is enabled.

For the chart itself (it's a problem with the other Gateway CRDs), I created https://github.com/kumahq/kuma/issues/3475

**Review by jpeach — APPROVED** [review]
This looks like a great start.

**Comment by jpeach** [comment]
> > I can live with GatewayInstance if you prefer that name
> 
> It's not my favorite name ever but we have time to change it if something better occurs to us.
> 
> > Should we always install this CRD, or only with some special flag?
> 
> I think we should always install it if gateway support is enabled.

That sounds reasonable.

File: deployments/charts/kuma/templates/cp-rbac.yaml [hunk 3]
```diff
@@ -129,6 +129,7 @@ rules:
     resources:
       - gateways
       - gatewayroutes
+      - gatewayinstances
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply deployments/charts/kuma/templates/cp-rbac.yaml:132]: For "now" perhaps, but we create them for Gateway API resources (which is part of `kuma-cp` at the moment)

File: pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_instance.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+// GatewayInstanceSpec specifies the options available for a GatewayDataplane.
+type GatewayInstanceSpec struct {
+	// Tags specifies the Kuma tags that are propagated to the managed
+	// dataplane proxies. These tags should include exactly one
+	// `kuma.io/service` tag, and should match exactly one Gateway
+	// resource.
+	//
+	// +required
+	// +kubebuilder:validation:MinLen=1
+	Tags map[string]string `json:"tags,omitempty"`
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_instance.go:32]: https://github.com/kumahq/kuma/issues/3502

File: deployments/charts/kuma/templates/cp-rbac.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply deployments/charts/kuma/templates/cp-rbac.yaml]: Removed all the RBAC and will put in the controller PR

File: deployments/charts/kuma/crds/gateway-instances.yaml [hunk 5]
```diff
@@ -0,0 +1,243 @@
+
```
[the remark this answers is not in the document]
Comment (jpeach) [reply deployments/charts/kuma/crds/gateway-instances.yaml:1]: Ah, I filed https://github.com/kumahq/kuma/issues/3504

**Pull request merged**